language_service/src/identifier_locator.rs

```diff
@@ -0,0 +1,318 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+
+use std::mem::replace;
+use std::rc::Rc;
+
+use crate::qsc_utils::{find_ident, find_item, span_contains, span_touches, Compilation};
+use qsc::ast::visit::{walk_expr, walk_namespace, walk_pat, walk_ty_def, Visitor};
+use qsc::{ast, hir, resolve};
+
+#[allow(unused_variables)]
+pub(crate) trait LocatorAPI<'package> {
+    fn at_callable_def(
+        &mut self,
+        context: &LocatorContext<'package>,
+        decl: &'package ast::CallableDecl,
```

Comment:
Symmetry:

I like the idea of all the `*_def` methods having the `Ident` node that contains the name as the first parameter (after `self` and `context` of course).

So they would all start with:

```rust
&mut self,
context: &LocatorContext<'package>,
name: &ast::Ident
```

I think that would make a very nice and clear pattern. The first parameter is consistently "the thing that the cursor is on."

So for `at_callable_def`, I think you should get `decl.name` in the locator and pass it in here explicitly. You're getting `decl.name` on all three implementations, so may as well do it in the locator.

```rs
 fn at_callable_def(
        &mut self,
        context: &LocatorContext<'package>,
        name: &'package ast::Ident,
        decl: &'package ast::CallableDecl,
    )
```

`at_field_def` and `at_new_type_def` already follow this pattern so great.

For `at_local_def`, you can reorder the parameters so that the name comes first just like the others:

```rs
    fn at_local_def(
        &mut self,
        context: &LocatorContext<'package>,
        ident: &'package ast::Ident,
        pat: &'package ast::Pat,
    )```

Reply:
done